File: components/signin/internal/identity_manager/account_capabilities_fetcher.cc
// Copyright 2021 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "components/signin/internal/identity_manager/account_capabilities_fetcher.h"

AccountCapabilitiesFetcher::AccountCapabilitiesFetcher(
    const CoreAccountInfo& account_info,
    FetchPriority fetch_priority,
    OnSomeCapabilitiesFetchedCallback on_some_capabilities_fetched_callback,
    OnAllFetchesCompleteCallback on_all_fetches_complete_callback)
    : account_info_(account_info),
      fetch_priority_(fetch_priority),
      on_some_capabilities_fetched_callback_(
          std::move(on_some_capabilities_fetched_callback)),
      on_all_fetches_complete_callback_(
          std::move(on_all_fetches_complete_callback)) {
  CHECK(on_some_capabilities_fetched_callback_);
  CHECK(on_all_fetches_complete_callback_);
}

AccountCapabilitiesFetcher::~AccountCapabilitiesFetcher() = default;

void AccountCapabilitiesFetcher::Start() {
  DCHECK(!started_);
  started_ = true;
  StartImpl();
}

void AccountCapabilitiesFetcher::UpdateAndCompleteFetchAndMaybeDestroySelf(
    const std::optional<AccountCapabilities>& capabilities) {
  if (capabilities.has_value()) {
    UpdateFetchedCapabilities(*capabilities);
  }
  CompleteFetchAndMaybeDestroySelf();
}

void AccountCapabilitiesFetcher::UpdateFetchedCapabilities(
    const AccountCapabilities& capabilities) {
  CHECK(on_some_capabilities_fetched_callback_);
  on_some_capabilities_fetched_callback_.Run(account_info_.account_id,
                                             capabilities);
}

void AccountCapabilitiesFetcher::CompleteFetchAndMaybeDestroySelf() {
  CHECK(on_all_fetches_complete_callback_);
  std::move(on_all_fetches_complete_callback_).Run(account_info_.account_id);
}

File: components/signin/public/android/java/src/org/chromium/components/signin/AccountCapabilitiesFetcher.java
// Copyright 2022 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.components.signin;

import org.jni_zero.CalledByNative;
import org.jni_zero.JniType;
import org.jni_zero.NativeMethods;

import org.chromium.build.annotations.NullMarked;
import org.chromium.components.signin.base.AccountCapabilities;
import org.chromium.components.signin.base.CoreAccountInfo;

/**
 * AccountCapabilitiesFetcher retrieves account capabilities from {@link AccountManagerFacade} and
 * passes them to native. It has a C++ counterpart located in
 * account_capabilities_fetcher_android.{h,cc}.
 */
@NullMarked
public class AccountCapabilitiesFetcher {
    private static final long INVALID_NATIVE_CALLBACK = 0;
    private final CoreAccountInfo mCoreAccountInfo;
    private long mNativeCallback;

    /**
     * Constructs a fetcher tied to a single account identified by {@code coreAccountInfo}.
     *
     * <p>WARNING: {@code nativeCallback} points to an unowned callback object allocated on native
     * heap. {@code this} *must* call {@link #onCapabilitiesFetchComplete} in order to clean up the
     * callback object properly.
     */
    @CalledByNative
    public AccountCapabilitiesFetcher(
            @JniType("CoreAccountInfo") CoreAccountInfo coreAccountInfo, long nativeCallback) {
        assert nativeCallback != INVALID_NATIVE_CALLBACK;
        mCoreAccountInfo = coreAccountInfo;
        mNativeCallback = nativeCallback;
    }
    ;

    @CalledByNative
    public void startFetchingAccountCapabilities() {
        AccountManagerFacadeProvider.getInstance()
                .getAccountCapabilities(mCoreAccountInfo)
                .then(
                        accountCapabilities -> {
                            onCapabilitiesFetchComplete(accountCapabilities);
                        });
    }

    private void onCapabilitiesFetchComplete(AccountCapabilities accountCapabilities) {
        assert mNativeCallback != INVALID_NATIVE_CALLBACK;
        AccountCapabilitiesFetcherJni.get()
                .onCapabilitiesFetchComplete(accountCapabilities, mNativeCallback);
        mNativeCallback = INVALID_NATIVE_CALLBACK;
    }

    @NativeMethods
    interface Natives {
        void onCapabilitiesFetchComplete(
                AccountCapabilities accountCapabilities, long nativeCallback);
    }
}
